app/services/cloud_front_request.rb

```diff
@@ -0,0 +1,61 @@
+class CloudFrontRequest
+  def initialize(request)
+    @request = request
+  end
+
+  def client_ip
+    addr = ip_and_port
+    return nil unless addr
+    addr[:ip]
+  end
+
+  def client_port
+    addr = ip_and_port
+    return nil unless addr
+    addr[:port]
+  end
+
+  # Source IP and port for client connection to CloudFront
+  def viewer_address
+    @request.get_header 'CloudFront-Viewer-Address'
+  end
+
+  # HTTP version used for client connection to CloudFront
+  def http_version
+    @request.get_header 'CloudFront-Viewer-Http-Version'
+  end
+
+  # TLS version and ciphers used for client connection to CloudFront
+  def tls_version
+    @request.get_header 'CloudFront-Viewer-TLS'
+  end
+
+  # ISO country code for IP client used to connect
+  def iso_country
+    @request.get_header 'CloudFront-Viewer-Country'
+  end
+
+  # ISO region subcode for IP client used to connect
+  def iso_region
+    @request.get_header 'CloudFront-Viewer-Country-Region'
+  end
+
+  private
+
+  # Regexp variables are not exactly intuitive:
+  # https://ruby-doc.org/core-2.5.1/Regexp.html#class-Regexp-label-Special+global+variables
+  def ip_and_port
+    return nil unless viewer_address
+    if viewer_address =~ /\[*\]:/ # IPv6
+      {
+        ip: "#{$`}]",
+        port: $',
```

Comment:
These are hideously ugly, hence me leaving in the link to the docs.

Reply:
The docs give us the cleaner names right there! LOL

```
$`      #=> "hay"
        # same as m.pre_match
$'      #=> "k"
        # same as m.post_match
```

so we can do:
```suggestion
    match = viewer_address.match(/\[*\]:/)

    if match
      {
        ip: match.pre_match,
        port: match.post_match,
```

Reply:
But on second thought, I think we don't need regexes at all, we can do this with split:

```ruby
*address_parts, port = viewer_address.split(':')
ip = address_pargs.join(':')
{
  ip: ip,
  port: port,
}
```

Reply:
D'oh! If only I had read down a few more lines in the example I would have found that. 🤦‍♂️😆 

I much prefer your approach here, though!